src/FeeRecipientDonationModule.sol

```diff
@@ -0,0 +1,423 @@
+    /// @return upkeepNeeded_ A boolean indicating whether an upkeep is to be performed.
+    /// @return performData_ The calldata to be passed to the upkeep function.
+    function checkUpkeep(bytes calldata checkData)
+        external
+        override
+        cannotExecute
+        returns (bool upkeepNeeded_, bytes memory performData_)
+    {
+        if (!SAFE.isModuleEnabled(address(this)) || !_isReady() || !_isValidSwapPath(swapPath)) {
+            // NOTE: explicit early return to checking rest of logic if these conditions are not met
+            return (upkeepNeeded_, performData_);
+        }
+
+        // total ebtc staked
+        uint256 storedTotalAssets = STAKED_EBTC.storedTotalAssets();
+        // total ebtc staked including left over rewards from the previous cycle
+        uint256 totalBalance = STAKED_EBTC.totalBalance();
+        uint256 residual = totalBalance - storedTotalAssets;
+        uint256 ebtcYield = storedTotalAssets * annualizedYieldBPS / (BPS * WEEKS_IN_YEAR);
+
+        if (residual >= ebtcYield) {
+            // there's still enough residual balance in the contract for this week
+            // performUpkeep still needs to be called to update lastProcessingTimestamp
+            // no need to claim additional PYS
+            return (true, abi.encode(0, 0));
+        }
+
+        uint256 ebtcAmountRequired = ebtcYield - residual;
+        uint256 stEthToClaim = ebtcAmountRequired * 1e18 / PRICE_FEED.fetchPrice();
+        uint256 collSharesToClaim = COLLATERAL.getSharesByPooledEth(stEthToClaim);
+        uint256 collSharesAvailable = _getFeeRecipientCollShares();
+
+        // cap by collSharesAvailable
+        if (collSharesToClaim > collSharesAvailable) {
+            collSharesToClaim = collSharesAvailable;
+
+            // recaclulate expected ebtcAmount after capping collSharesToClaim
+            ebtcAmountRequired = COLLATERAL.getPooledEthByShares(collSharesToClaim) * PRICE_FEED.fetchPrice() / 1e18;
+        }
+
+        return (true, abi.encode(collSharesToClaim, ebtcAmountRequired));
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    // INTERNAL
+    ////////////////////////////////////////////////////////////////////////////
+
+    function _isReady() private view returns (bool) {
+        return lastProcessingTimestamp < _lastRewardCycle();
+    }
+
+    function _isValidSwapPath(bytes memory _swapPath) private returns (bool) {
+        (uint256 spotPrice, , , ) = QUOTER.quoteExactInput(_swapPath, wstETH.getWstETHByStETH(1e18));
+        uint256 oraclePrice = PRICE_FEED.fetchPrice();
+
+        uint256 absDiff;
+        if (spotPrice > oraclePrice) {
+            absDiff = spotPrice - oraclePrice;
+        } else if (spotPrice < oraclePrice) {
+            absDiff = oraclePrice - spotPrice;
+        }
+
+        // make sure price diff % never exceeds minOut %
+        return (absDiff * BPS / oraclePrice) <= (BPS - minOutBPS);
```

Comment:
# Hunch - This looks off

We're checking agains 1e18 instead of the amount we want to sell
We're checking even when we get positive slippage

I think this could be simplified to just check if UniV3 will offer subpar exec vs the oracle
And for having better execution than the Orace due to drift that would have to be monitored